README
# ads-b-collect

# (c) Kazansky137 - Mon Aug 24 20:01:20 CEST 2020

Credits and technical details follows later in this file.
I would like to emphasize on this software on a practical
point of view, I mean how to use it.

Manual
------

Collect and analyze ADS-B messages and related
----------------------------------------------

This work is greatly based on the library "pyModeS" of Junzi Sun.

The 1090MHz Riddle - An open-access book about decoding Mode-S and ADS-B data
	- https://mode-s.org/decode/
    - https://github.com/junzis/the-1090mhz-riddle

The Python Decoder for ADS-B (DF17) and Mode-S Comm-B (DF20/21)
	- https://github.com/junzis/pyModeS

pyModeS: Decoding Mode-S Surveillance Data for Open Air Transportation Research
    - http://tiny.cc/19j1dz
    - @article{sun2019pymodes,
      author={J. {Sun} and H. {V\^u} and J. {Ellerbroek} and J. M. {Hoekstra}},
      journal={IEEE Transactions on Intelligent Transportation Systems},
      title={pyModeS: Decoding Mode-S Surveillance Data for
             Open Air Transportation Research},
      year={2019},
      doi={10.1109/TITS.2019.2914770},
      ISSN={1524-9050},
      }

ADS-B messages used in this project will be collected from sdfr[xy]
Raspberry Pi's in the slashdev.be network :

- sdfr241 - gbx-se : sdfr241-w.slashdev.be
- sdfr242 - gbx-nw : sdfr242.slashdev.be
- sdfr243 - lln-se : sheremetyevo.dyndns.org
- sdfr244 - lln-nw : domodedovo.dyndns.org

On each Raspberry, the dump1090-fa process is collecting raw ADS-B
messages from USB 1090 MHz receivers ; "fa" stands from FligtAware
as there should be different implementations of the dump1090 process.

dump1090-fa open ports to read/write ADS-B messages in different format :
- 30001 -     : TCP raw input listen port
- 30002 -     : TCP raw output listen port
- 30003 - sbs : TCP BaseStation output listen port
- 30004 - bi  : TCP Beast input listen port
- 30005 - bo  : TCP Beast output listen port


HISTORY
# ads-b-collect

# (c) Kazansky137 - Tue Aug 25 20:28:24 UTC 2020

2019 Dec 07 22:53 : master - 0026e17 : Initial release
2019 Dec 10 22:43 : devout - 3e66357 : Improved colored output
2019 Dec 11 11:42 : devout - 26b700e : Bugs correction
2019 Dec 11 17:27 : devout - 29e353e : Replaced config/alerts.template by alerts
2019 Dec 12 12:48 : devout - fd13d23 : Updated log output with ICAO, add rates
2019 Dec 13 22:42 : devout - 141c34d : Updated log output, pep8 compliant
2019 Dec 14 18:17 : devout - 06189df : Updated log output, Added run-file.sh
2019 Dec 17 16:48 : devout - bd4cf84 : Changed emg to urg. Increased cnt period
2019 Dec 17 16:57 : devout - a7fadef : Updated config/alerts.txt file
2019 Dec 17 21:19 : devout - 3404135 : Moved log alert from flights-2 to alert
2019 Dec 17 22:14 : devout - 4dfa825 : Added ringring feature to send SMS
2019 Dec 18 17:04 : devout - 8c4a917 : Integrated ringring to alert.py
2019 Dec 19 15:18 : devout - 10a7c92 : Added icao-codes to add tail n° in SMS.
2020 Jan 09 22:26 : devout - 0d8a297 : Updated config/icao-tail.txt.
2020 Jan 13 19:44 : devout - 422ccfa : Added email alerting.
2020 Jan 13 19:59 : devout - 4909ee1 : Moved ringring from flights-2 to alert
2020 Mar 15 19:23 : devout - 288f8d5 : Updated config/icao-tail.txt.
2020 Mar 16 22:41 : master - b5c3973 : Now decode msgs from fr24 dump1090
2020 Mar 17 19:03 : master - 8b402b1 : Made pep8 v2.3.1 compliant.
2020 Mar 18 19:27 : devpos - 1d6c312 : Moved raw statistics to Discover
2020 Mar 24 00:42 : devpos - 01f99c5 : Added stats for crc, dfmt, tc28...
2020 Mar 27 20:00 : devpos - 399d46c : Moved "bad" message in discover.py
2020 Apr 03 22:15 : devpos - 858d651 : Moved catxt to common message return dict
2020 Apr 03 23:30 : devpos - 5311d32 : Corrected ca() handling + added stats.
2020 Apr 04 19:12 : devpos - c1042ff : Get all ADS-B data from discover
2020 Apr 04 22:40 : master - 14b5a25 : Updated config/icao-tail.txt.
2020 Apr 04 23:14 : master - 6cdb593 : Added TREE history file.
2020 Apr 08 22:40 : devpos - 3822811 : Added management of altitude.
2020 Apr 09 23:03 : devpos - e5144e8 : Bugs corrected + add ls, fs, min, max alt
2020 Apr 11 23:15 : devpos - 8f13953 : Adapted flights output. Added alt filter
2020 Apr 12 16:55 : devpos - da20343 : Adapted altitude filter
2020 Apr 12 17:38 : devpos - b7a86e0 : Made pep8 v2.3.1 compliant.
2020 Apr 14 17:35 : devpos - 3f5da30 : Alerts now handle altitude
2020 Apr 15 17:32 : master - 56617ae : Updated HISTORY and TREE files
2020 Apr 16 20:04 : master - c278ff5 : Merge branch 'devpos' @3f5da30
2020 Apr 17 19;06 : devpos - 4ba890c : Simplified import of pyModeS
2020 Apr 18 16:37 : devpos - ee7ea20 : Now handle baroc and gnss alt + bug corr
2020 Apr 18 22:07 : devpos - 44a31c7 : Now handle 56 bits messages
2020 Apr 21 19:00 : devpos - d319ad5 : Manage lat and long, added scripts
2020 Apr 21 21:27 : devpos - 1c65a9c : Print flightlist is last seen order
2020 Apr 22 12:08 : devpos - 07b0e68 : Bugs corr, ordered flight list printout
2020 Apr 22 20;04 : devpos - e5b84e5 : Compute and print distance and bearing
2020 Apr 23 00;09 : devpos - a8958bb : Merged config files to config.txt
2020 Apr 24 19:18 : devpos - e89f4b0 : Alert based on Flights, improved outputs
2020 Apr 24 19:28 : master - 25a14dc : Updated HISTORY and TREE files
2020 Apr 25 17:38 : devpos - 486ec64 : Bug corr in sh., skip null icao msgs
2020 Apr 25 18:05 : master - 6ed91f1 : Merge branch 'devpos' @486ec64
2020 Apr 25 23:41 : devpos - 1704580 : Manage velocity speed, heading, rate
2020 Apr 26 17:26 : master - 2d36c39 : Updated HISTORY and TREE files
2020 Apr 26 17:56 : master - ded6657 : Merge branch 'devpos' @1704580
2020 Apr 26 18:50 : master - 226ad12 : Updated HISTORY and TREE files
2020 Apr 26 20:29 : devout - ecdc627 : Moved src details to config.txt
2020 Apr 29 22:21 : devout - 1a7a5a4 : Discover is now it's own process
2020 May 01 22:34 : devout - 3edf8a2 : Improved mgmt of exceptions & check_msg
2020 May 02 19:10 : devout - 7fb838e : Distance and bearing from pyModeS aero
2020 May 02 19:25 : master - b2e0250 : Merge branch 'devout' @7fb838e
2020 May 03 21:40 : devout - f101c5b : Handle args sms, added run-legacy.sh
2020 May 05 22:38 : devout - a4962b9 : Updated handling of exceptions + --debug
2020 May 05 22:48 : master - 69e2d78 : Merge branch 'devout' @a4962b9
2020 May 09 23:27 : devout - 6cb253a : Added magic typed legacy files
2020 May 09 23:41 : master - 00bd8d2 : Merge branch 'devout' @6cb253a
2020 May 10 19:58 : devout - 7b502d6 : Use of magic characters in collect/disc.
2020 May 11 22:45 : devout - 6cd7616 : Updated run-file.sh to run legacy files
2020 May 11 23:01 ; master - 0da0578 : Merge branch 'devout' @6cd7616
2020 May 12 23:04 ; devout - 4358121 : Added args, mail alerts live
2020 May 14 19:29 ; devout - 019e5fb : Added args, mgmt exceptions, bad perf.
2020 May 17 19:56 ; devout - 2dbbe67 : Bad perf corrected in tsmessage.py
2020 May 19 19:33 : devout - 65cf456 : Improved performances for debug.
2020 May 19 19:49 : master - 4c6c885 : Merge branch 'devout' @65cf456
2020 May 28 06:21 : devprf - 3a21246 : Improved args mgmt, added profiling
2020 May 28 22:57 : devprf - 700702c : Improved load_config and _debug handling
2020 Jun 04 22:49 : devprf - d55b595 : Improved args, exceptions + filter planes
2020 Jun 14 17:42 : master - 53a8a3e : Added aircrafts db download.
2020 Aug 24 20:01 : master - 2205fd4 : Use last-log, removed mail opt. Added v. %MBR24-3.1
2020 Aug 25 19:43 : devout - 5606ce6 : Merge branch 'master' @2205fd4
2020 Aug 25 22:17 : devout - b7cb74e : Bug corrected with xterm font 9x15
2020 Aug 25 ..:.. : master - ....... : Merge branch 'devout' @b7cb74e


discover.py
#! /usr/bin/env python3

# (c) Kazansky137 - Mon Aug 24 20:01:20 CEST 2020

import sys
import os
from time import time
from common import log, adsb_ca, load_config
from pyModeS import common, adsb
from tsmessage import TsMessage
import traceback
import signal
import re
from common import print_config

_debug = 0

ca_msg = ["None",   # 0 - No ADSB-Emitter
          "Light",  # 1 - Light < 15500 lbs
          "Small",  # 2 - Small   15500 to  75000 lbs
          "Large",  # 3 - 75000 to 300000 lbs
          "HVort",  # 4 - High Vortex Large (aircraft such as B-757)
          "Heavy",  # 5 - Heavy > 300000 lbs
          "HPerf",  # 6 - High Performance > 5 g acceleration and > 400 kts
          "Rotor"   # 7 - Rotorcraft
          ]

df_msg = ["ADS-B 56 bits Short air to air ACAS BDS 0.E",    # 00
          "",                                               # 01
          "",                                               # 02
          "",                                               # 03
          "ADS-B 56 bits Surveillance altitude",            # 04
          "ADS-B 56 bits Surveillance identity",            # 05
          "",                                               # 06
          "",                                               # 07
          "",                                               # 08
          "",                                               # 09
          "",                                               # 10
          "ADS-B 56 bits Mode-S only all call reply",       # 11
          "",                                               # 12
          "",                                               # 13
          "",                                               # 14
          "",                                               # 15
          "",                                               # 16
          "ADS-B ES (112 bits) BDS 0.[5..9]",               # 17
          "",                                               # 18
          "",                                               # 19
          "Mode-S EHS altitude reply BDS [4,5,6].0",        # 20
          "Mode-S EHS identity reply BDS [4,5,6].0",        # 21
          "",                                               # 22
          "",                                               # 23
          "",                                               # 24
          "",                                               # 25
          "",                                               # 26
          "",                                               # 27
          "",                                               # 28
          "",                                               # 29
          "",                                               # 30
          ""                                                # 31
          ]

tc_msg = ["Aircraft identification",                   # 01
          "Aircraft identification",                   # 02
          "Aircraft identification",                   # 03
          "Aircraft identification",                   # 04
          "Surface position",                          # 05
          "Surface position",                          # 06
          "Surface position",                          # 07
          "Surface position",                          # 08
          "Airborne position (w/ Baro Altitude)",      # 09
          "Airborne position (w/ Baro Altitude)",      # 10
          "Airborne position (w/ Baro Altitude)",      # 11
          "Airborne position (w/ Baro Altitude)",      # 12
          "Airborne position (w/ Baro Altitude)",      # 13
          "Airborne position (w/ Baro Altitude)",      # 14
          "Airborne position (w/ Baro Altitude)",      # 15
          "Airborne position (w/ Baro Altitude)",      # 16
          "Airborne position (w/ Baro Altitude)",      # 17
          "Airborne position (w/ Baro Altitude)",      # 18
          "Airborne velocities",                       # 19
          "Airborne position (w/ GNSS Height)",        # 20
          "Airborne position (w/ GNSS Height)",        # 21
          "Airborne position (w/ GNSS Height)",        # 22
          "Reserved",                                  # 23
          "Reserved",                                  # 24
          "Reserved",                                  # 25
          "Reserved",                                  # 26
          "Reserved",                                  # 27
          "Aircraft status",                           # 28
          "Target state and status information",       # 29
          "",                                          # 30
          "Aircraft operation status"                  # 31
          ]


class Discover:
    def handler_hup(self, _signum, _frame):
        self.signal_hup = 1

    def __init__(self):
        self.params = {}
        load_config(self.params, "config/config.txt")

        global _debug
        _debug = 1 if self.params["arg_debug"] else 0

        self.msgs_discovered = 0

        self.msgs_curr_rread = 0
        self.msgs_curr_total = 0
        self.msgs_curr_len28 = 0
        self.msgs_curr_short = 0

        self.msgs_last_rread = 0
        self.msgs_last_total = 0
        self.msgs_last_len28 = 0
        self.msgs_last_short = 0

        self.df = [0] * 32          # Downlink Format : 5 bits  1 ..  5
        self.ca = [0] *  8          # Capability      : 3 bits  6 ..  8
        self.tc = [0] * 32          # Type Code       : 5 bits 33 .. 37

        self.parity_check_ok = 0
        self.parity_check_ko = 0

        self.exc_unavailable = 0
        self.exc_missing     = 0
        self.exc_velocity    = 0
        self.exc_heading     = 0
        self.exc_rocd        = 0
        self.exc_crc_ko      = 0
        self.exc_other       = 0

        # Time counters
        self.t_curr = time()
        self.t_last = time()

        # Signals handling
        self.signal_hup = 0
        signal.signal(signal.SIGHUP, self.handler_hup)

    def check_msg(self, msg):
        df = common.df(msg)
        msglen = len(msg)
        if df == 17 and msglen == 28:
            if common.crc(msg) == 0:
                return True
        elif df in [20, 21] and msglen == 28:
            return True
        elif df in [4, 5, 11] and msglen == 14:
            return True
        else:
            return False

    def ca_txt(self, index):
        return ca_msg[index]

    def message(self, msg):
        # Printout of statistics
        if self.signal_hup == 1:
            self.logstats()
            self.signal_hup = 0

        ret_dict = {}
        ret_dict['ret'] = 0
        ret_dict['type'] = ""

        self.msgs_curr_total = self.msgs_curr_total + 1

        if len(msg) == 26 or len(msg) == 40:
            # Some version of dump1090 have the 12 first characters used w/
            # some date (timestamp ?). E.g. sdbr245 feeding flightradar24.
            # Strip 12 first characters.
            msg = msg[12:]

        if len(msg) < 28:        # Message length 112 bits
            self.msgs_curr_short = self.msgs_curr_short + 1
        else:
            self.msgs_curr_len28 = self.msgs_curr_len28 + 1

        ret_dict['crc'] = self.check_msg(msg)
        if ret_dict['crc']:
            self.parity_check_ok = self.parity_check_ok + 1
        else:
            self.parity_check_ko = self.parity_check_ko + 1

        # Do not manage messages with bad CRC
        if ret_dict['crc'] is not True:
            raise ValueError("CrcKO")

        dfmt = common.df(msg)
        ret_dict['dfmt'] = dfmt
        self.df[dfmt] = self.df[dfmt] + 1

        ret_dict['ic'] = common.icao(msg)

        if dfmt in [17, 18]:    # Downlink format 17 or 18
            tc = common.typecode(msg)
            ret_dict['tc'] = tc
            self.tc[tc] = self.tc[tc] + 1

            lat_ref = float(self.params["lat"])
            long_ref = float(self.params["long"])

            if tc == 4:         # Aircraft identification
                self.msgs_discovered = self.msgs_discovered + 1
                ret_dict['type'] = "CS"
                ret_dict['cs'] = adsb.callsign(msg)
                ca = adsb_ca(msg)
                ret_dict['ca'] = ca_msg[ca]
                self.ca[ca] = self.ca[ca] + 1
            elif 9 <= tc <= 18:
                self.msgs_discovered = self.msgs_discovered + 1
                ret_dict['type'] = "LB"
                ret_dict['altb'] = adsb.altitude(msg)
                (lat, long) = adsb.position_with_ref(msg, lat_ref, long_ref)
                ret_dict['lat'] = lat
                ret_dict['long'] = long
            elif tc == 19:
                self.msgs_discovered = self.msgs_discovered + 1
                ret_dict['type'] = "VH"
                _dict = adsb.velocity(msg)
                if _dict is None:
                    raise ValueError("AdsbVelocity")
                (ret_dict['speed'], ret_dict['head'],
                 ret_dict['rocd'], var) = _dict
                if ret_dict['head'] is None:
                    raise ValueError("AdsbHeading")
                if ret_dict['rocd'] is None:
                    raise ValueError("AdsbRocd")
            elif 20 <= tc <= 22:
                self.msgs_discovered = self.msgs_discovered + 1
                ret_dict['type'] = "LG"
                ret_dict['altg'] = adsb.altitude(msg)
                (lat, long) = adsb.position_with_ref(msg, lat_ref, long_ref)
                ret_dict['lat'] = lat
                ret_dict['long'] = long
        elif dfmt in [5, 21]:
                self.msgs_discovered = self.msgs_discovered + 1
                ret_dict['type'] = "SQ"
                ret_dict['sq'] = common.idcode(msg)

        if dfmt in [0, 4, 16, 20]:
            self.msgs_discovered = self.msgs_discovered + 1
            ret_dict['type'] = "AL"
            _alt = common.altcode(msg)
            alt = _alt if _alt is not None else 0
            ret_dict['alt'] = alt

        return ret_dict

    def logstats(self):
        # Time
        self.t_last = self.t_curr
        self.t_curr = time()
        self.t_diff = self.t_curr - self.t_last

        # Messages
        delta_rread = self.msgs_curr_rread - self.msgs_last_rread
        delta_total = self.msgs_curr_total - self.msgs_last_total
        delta_len28 = self.msgs_curr_len28 - self.msgs_last_len28
        delta_short = self.msgs_curr_short - self.msgs_last_short

        delta_rread = int(delta_rread / self.t_diff)
        delta_total = int(delta_total / self.t_diff)
        delta_len28 = int(delta_len28 / self.t_diff)
        delta_short = int(delta_short / self.t_diff)

        self.msgs_last_rread = self.msgs_curr_rread
        self.msgs_last_total = self.msgs_curr_total
        self.msgs_last_len28 = self.msgs_curr_len28
        self.msgs_last_short = self.msgs_curr_short

        log("Running   : Raw Read  {:>12,d} raw read ({:5d} /s)"
            .format(self.msgs_last_rread, int(delta_rread)))
        log("Running   : Raw Read  {:>12,d} messages ({:5d} /s)"
            .format(self.msgs_last_total, int(delta_total)))
        log("Running   : Raw Read  {:>12,d} msglen28 ({:5d} /s)"
            .format(self.msgs_last_len28, int(delta_len28)))
        log("Running   : Raw Read  {:>12,d} msgshort ({:5d} /s)"
            .format(self.msgs_last_short, int(delta_short)))

        log("Running   : Raw Read  {:>12,d} discovered"
            .format(self.msgs_discovered))

        log("Running   : Raw Read  {:>12,d} check ok"
            .format(self.parity_check_ok))
        log("Running   : Raw Read  {:>12,d} check ko"
            .format(self.parity_check_ko))

        log("Running   : Raw Read  {:>12,d} exceptions resource unavailable"
            .format(self.exc_unavailable))
        log("Running   : Raw Read  {:>12,d} exceptions missing message"
            .format(self.exc_missing))
        log("Running   : Raw Read  {:>12,d} exceptions missing velocity      *"
            .format(self.exc_velocity))
        log("Running   : Raw Read  {:>12,d} exceptions missing heading       *"
            .format(self.exc_heading))
        log("Running   : Raw Read  {:>12,d} exceptions missing rocd          *"
            .format(self.exc_rocd))
        log("Running   : Raw Read  {:>12,d} exceptions crc ko                *"
            .format(self.exc_crc_ko))
        log("Running   : Raw Read  {:>12,d} exceptions other                 ?"
            .format(self.exc_other))

        s = 0
        for i in range(len(self.df)):
            v = self.df[i]
            if v > 0:
                s = s + v
                log("Running   : Raw Read  {:>12,d} dfmt[{:2d}] {:s}"
                    .format(v, i, df_msg[i]))
        log("Running   : Raw Read  {:>12,d} df cnt".format(s))

        s = 0
        for i in range(len(self.tc)):
            v = self.tc[i]
            if v > 0:
                s = s + v
                log("Running   : Raw Read  {:>12,d} tc17[{:2d}] {:s}"
                    .format(v, i, tc_msg[i-1]))
        log("Running   : Raw Read  {:>12,d} tc17/18 cnt".format(s))

        s = 0
        for i in range(len(self.ca)):
            v = self.ca[i]
            if v > 0:
                s = s + v
                log("Running   : Raw Read  {:>12,d} ca[{:2d}] {:s}"
                    .format(v, i, ca_msg[i]))
        log("Running   : Raw Read  {:>12,d} ca cnt".format(s))


if __name__ == "__main__":
    log("Running: Pid {:5d}".format(os.getpid()))

    disc = Discover()
    cnt = 0
    first_line = True

    for line in sys.stdin:
        if first_line:
            first_line = False
            regex = re.compile('(%MBR24-[1-3].[0-9]) ([A-Z\-]+)$')
            result = regex.match(line)
            if result is None:
                raise ValueError("Invalid magic characters")
            if _debug:
                log("debug first line:", result.group(1), result.group(2))
            disc.params['in_vers'] = result.group(1)
            disc.params['in_name'] = result.group(2)
            if disc.params['in_vers'] == "%MBR24-3.1":
                word_one = 0
                word_two = 1
            elif disc.params['in_vers'] == "%MBR24-3.0":
                word_one = 0
                word_two = 1
            elif disc.params['in_vers'] == "%MBR24-2.0":
                word_one = 1
                word_two = 2
            elif disc.params['in_vers'] == "%MBR24-1.0":
                word_one = 0
                word_two = 1
            else:
                raise ValueError("Invalid input version")
            continue

        if _debug:
            log("debug read:", line, end='')

        if line[0] == '#':
            continue

        cnt = cnt + 1
        disc.msgs_curr_rread = disc.msgs_curr_rread + 1

        words = line.split()
        try:
            ts_msg = TsMessage(words[word_one], words[word_two])
            ret_dict = disc.message(words[word_two])
            if ret_dict['ret'] == 0:
                ts_msg.disc(ret_dict)
                # ts_msg.print()
                ts_msg.print_legacy()

        except IndexError as e:
            if len(words) == 1:
                disc.exc_missing = disc.exc_missing + 1
                if _debug:
                    log("debug exception: line {:10d}: Missing message"
                        .format(cnt+1))
            else:
                log("Exception: line {:10d}: {:s} {:s}".
                    format(cnt, str(type(e)), str(e)))
        except ValueError as e:
            if words[0] in ["Unexpected", "Error:"]:
                disc.exc_unavailable = disc.exc_unavailable + 1
                if _debug:
                    log("debug exception: line {:10d}: Resource unavailable"
                        .format(cnt+1))
            elif str(e) == "AdsbVelocity":
                disc.exc_velocity = disc.exc_velocity + 1
                log("Exception: line {:10d}: Adsb velocity none".format(cnt+1))
            elif str(e) == "AdsbHeading":
                disc.exc_heading = disc.exc_heading + 1
                log("Exception: line {:10d}: Adsb heading none".format(cnt+1))
            elif str(e) == "AdsbRocd":
                disc.exc_rocd = disc.exc_rocd + 1
                log("Exception: line {:10d}: Adsb rocd none".format(cnt+1))
            elif str(e) == "CrcKO":
                disc.exc_crc_ko = disc.exc_crc_ko + 1
                if _debug:
                    log("debug exception: line {:10d}: CRC check failure"
                        .format(cnt+1))
            else:
                log("Exception: line {:10d}: {:s} {:s}".
                    format(cnt+1, str(type(e)), str(e)))
        except Exception as e:
            disc.exc_other = disc.exc_other + 1
            log("Exception: line {:10d}: {:s} {:s}".
                format(cnt+1, str(type(e)), str(e)))
            log(traceback.format_exc())

    disc.logstats()

    sys.exit(0)
